yt/frontends/amrvac/fields.py

```diff
@@ -45,33 +74,65 @@ class AMRVACFieldInfo(FieldInfoContainer):
         ("b1", ("code_magnetic", ["magnetic_1"], None)),
         ("b2", ("code_magnetic", ["magnetic_2"], None)),
         ("b3", ("code_magnetic", ["magnetic_3"], None))
+        )) + known_dust_fields
+        # in python3, there is no need for this tuple+list conversion, it suffices to write
+        # known_other_fields = (..., *known_dust_fields)
     )
 
     known_particle_fields = ()
 
-    def setup_fluid_fields(self):
-        def _v1(field, data):
-            return data["gas", "moment_1"] / data["gas", "density"]
-        def _v2(field, data):
-            return data["gas", "moment_2"] / data["gas", "density"]
-        def _v3(field, data):
-            return data["gas", "moment_3"] / data["gas", "density"]
+    def _create_velocity_fields(self, idust=None):
+        if idust is None:
+            dust_flag = dust_label = ""
+        else:
+            dust_flag = "d%d" % idust
+            dust_label = "dust%d_" % idust
 
         us = self.ds.unit_system
-        aliases = direction_aliases[self.ds.geometry]
-        for idir, alias, func in zip("123", aliases, (_v1, _v2, _v3)):
-            if not ("amrvac", "m%s" % idir) in self.field_list:
+        for i_, alias in enumerate(direction_aliases[self.ds.geometry]):
+            idir = i_ + 1 # direction index starts at 1 in AMRVAC
+            if not ("amrvac", "m%d%s" % (idir, dust_flag)) in self.field_list:
                 break
-            self.add_field(("gas", "velocity_%s" % alias), function=func,
-                            units=us['velocity'],
-                            dimensions=dimensions.velocity,
-                            sampling_type="cell")
-            self.alias(("gas", "velocity_%s" % idir), ("gas", "velocity_%s" % alias),
+            self.add_field(("gas", "%svelocity_%s" % (dust_label, alias)), function=_velocity(idir, idust),
+                            units=us['velocity'], dimensions=dimensions.velocity, sampling_type="cell")
+            self.alias(("gas", "%svelocity_%d" % (dust_label, idir)), ("gas", "%svelocity_%s" % (dust_label, alias)),
                         units=us["velocity"])
-            self.alias(("gas", "moment_%s" % alias), ("gas", "moment_%s" % idir),
+            self.alias(("gas", "%smoment_%s" % (dust_label, alias)), ("gas", "%smoment_%d" % (dust_label, idir)),
                         units=us["density"]*us["velocity"])
```

Comment:
At this point we're really starting to intertwine dust fields and regular fields, making it very hard to distinguish between the two. I see why you're doing it this way (it's nice and short), but I'm wondering if isn't more convenient to calculate dust fields in a completely separate method (`_setup_dust_fluid_fields()` or something like that), and call it _only_ when there is dust present? It will make it easier for maintaining and/or extending the fields specifically for each case (dust or no dust, or others in the future), no?

Reply:
Actually I realised that in the process but decided against making it a seperate case because I don't think it hurts to have  "total_dust_density" or "dust_to_gas_ratio" recipes even is cases where both are 0 (no dust). Even though, this would create a de facto inconsistency with how we treat e.g. magnetic energy... I'll think I'll change it then.

Reply:
Actually I realised that in the process but decided against making it a seperate case because I don't think it hurts to have "total_dust_density" or "dust_to_gas_ratio" recipes even is cases where both are 0 (no dust). Even though, this would create a de facto inconsistency with how we treat e.g. magnetic energy... I'll think I'll change it then.

Reply:
In any case I'm against duplicated code when I know how to avoid it (most of the time), so I don't think it's a good strategy to seperate them more than they are.